prog/dftb+/lib_timedep/pprpa.F90

```diff
@@ -0,0 +1,716 @@
+    real(dp), intent(in) :: SSqr(:,:)
+
+    !> chemical species of each atom
+    integer, intent(in) :: species0(:)
+
+    !> real number of electrons in system
+    real(dp), intent(in) :: rnel
+
+    !> Atomic neighbour lists
+    integer, intent(in) :: iNeighbour(0:,:)
+
+    !> Mapping of atom number to central cell atom number
+    integer, intent(in) :: img2CentCell(:)
+
+    !> data type for atomic orbital information
+    type(TOrbitals), intent(in) :: orb
+
+    !> print tag information
+    logical, intent(in) :: tWriteTagged
+
+    !> File name for regression data
+    character(*), intent(in) :: autotestTag
+
+    !> Tagged writer
+    type(TTaggedWriter), intent(inout) :: taggedWriter
+
+    !> virtual orbital constraint?
+    logical, intent(in) :: tConst
+
+    !> number of virtual orbitals
+    integer, intent(in) :: nConst
+
+
+    logical :: tSpin
+
+    real(dp), allocatable :: stimc(:,:,:)
+    real(dp), allocatable :: gamma_eri(:,:)
+
+    character, allocatable :: symmetries(:)
+
+    !> file handle for excitation energies
+    integer:: fdExc
+
+    integer, allocatable :: iAtomStart(:)
+    integer :: nocc, nvir, nxoo, nxvv, dim_rpa
+    integer :: norb, natom
+    integer :: iSpin, isym
+    integer :: nSpin
+    integer :: fdTagged
+    character :: sym
+
+    integer :: homo = 0
+    real(dp) :: eval_0 = 0.0_dp
+
+    !> ppRPA eigenvalues (two-electron addition/removal energies)
+    real(dp), allocatable :: pp_eval(:)
+
+    !> ppRPA eigenvector
+    real(dp), allocatable :: vr(:,:)
+
+
+    @:ASSERT(fdExc > 0)
+
+    nSpin = size(grndEigVal, dim=2)
+    @:ASSERT(nSpin > 0 .and. nSpin <=2)
+
+    norb = orb%nOrb
+    natom = size(species0)
+    tSpin = (nSpin == 2)
+
+    ALLOCATE(iAtomStart(size(denseDesc%iAtomStart)))
+    iAtomStart = denseDesc%iAtomStart
+
+
+    ! Select symmetries to process
+    if (.not. tSpin) then
+      select case (symc)
+      case ("B")
+        ALLOCATE(symmetries(2))
+        symmetries(:) = [ "S", "T" ]
+      case ("S")
+        ALLOCATE(symmetries(1))
+        symmetries(:) = [ "S" ]
+      case ("T")
+        ! like B. Triplet calculation requires first singlet exc. energy
+        ALLOCATE(symmetries(2))
+        symmetries(:) = [ "S", "T" ]
+      end select
+    else
+      call error("spin-unrestricted calculations currently not possible with pp-RPA ")
+    end if
+
+    ! Allocation for general arrays
+    ALLOCATE(gamma_eri(natom, natom))
+    ALLOCATE(stimc(norb, norb, nSpin))
+
+    if (tWriteTagged) then
+      open(newUnit=fdTagged, file=autotestTag, position="append")
+    end if
+
+    ! Overlap times wave function coefficients - most routines in DFTB+ use lower triangle (would
+    ! remove the need to symmetrize the overlap and ground state density matrix in the main code if
+    ! this could be used everywhere in these routines)
+    do iSpin = 1, nSpin
+      call symm(stimc(:,:,iSpin), "L", SSqr, grndEigVecs(:,:,iSpin))
+    end do
+
+    nocc = nint(rnel) / 2
```

Comment:
Needs a test that this really is a closed shell system

Reply:
Anything other than +2 and spin free then should give warnings (call warn()), but if its useful to allow other cases the code should not stop.